# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/gtk/WaylandBuffer.cpp
# Commit: 9b286c5c32d1
# Full Hash: 9b286c5c32d1accc0025872508b34a1eeb0d2d64
# Author: stransky <stransky@redhat.com>
# Date: 2025-07-22 04:26:57
# Regressor Bug: 1978207
# File Overlap Count: 4
# Description:
#   Bug 1978207 [Wayland] Remove detached WaylandBuffer transactions from WaylandSurface r=emilio
#   
#   When WaylandBuffer transaction is detached from compositor, remove it from WaylandSurface.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D257976
# ==============================================================================

diff -r d73d4fbe1330 -r 9b286c5c32d1 widget/gtk/WaylandBuffer.cpp
--- a/widget/gtk/WaylandBuffer.cpp	Mon Jul 21 15:15:52 2025 +0000
+++ b/widget/gtk/WaylandBuffer.cpp	Mon Jul 21 15:19:40 2025 +0000
@@ -350,6 +350,8 @@
         [](void* aData, wl_buffer* aBuffer) {
           auto transaction = static_cast<BufferTransaction*>(aData);
           if (transaction) {
+            // BufferDetachCallback can delete transaction
+            RefPtr grip{transaction};
             transaction->BufferDetachCallback();
           }
         }};
@@ -363,13 +365,18 @@
 }
 
 void BufferTransaction::BufferDetachCallback() {
-  LOGWAYLAND("BufferTransaction::BufferDetach() [%p] WaylandBuffer [%p] ", this,
-             (void*)mBuffer);
   WaylandSurfaceLock lock(mSurface);
+  LOGWAYLAND(
+      "BufferTransaction::BufferDetach() [%p] WaylandBuffer [%p] attached to "
+      "WaylandSurface %d",
+      this, (void*)mBuffer, mSurface->IsBufferAttached(mBuffer));
+
   if (mBufferState != BufferState::WaitingForDelete) {
     mBufferState = BufferState::Detached;
 
-    if (mIsExternalBuffer) {
+    // Delete this transaction if WaylandSurface uses different WaylandBuffer
+    // already, we don't need to keep it for recycling.
+    if (!mSurface->IsBufferAttached(mBuffer)) {
       DeleteTransactionLocked(lock);
     }
   }
@@ -451,6 +458,10 @@
   LOGWAYLAND("BufferTransaction::DeleteLocked() [%p] WaylandBuffer [%p]", this,
              (void*)mBuffer);
   MOZ_DIAGNOSTIC_ASSERT(mBufferState == BufferState::Deleted);
+  MOZ_DIAGNOSTIC_ASSERT(mSurface);
+
+  // Unlink from Surface
+  mSurface->RemoveTransactionLocked(aSurfaceLock, this);
   mSurface = nullptr;
 
   // This can destroy us